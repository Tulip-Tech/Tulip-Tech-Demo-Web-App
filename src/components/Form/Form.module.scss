@use 'src/styles/variables';
@use 'src/styles/theme';
@use 'src/styles/mixins/responsive';

button {
  margin-right: variables.$base-spacing;
}

.flexBox {
  display: flex;
  flex-direction: column;
  > * {
    margin-bottom: calc(#{variables.$base-spacing} / 2);
  }
}

.controls {
  margin-top: variables.$base-spacing;

  > button {
    margin-right: 8px;
  }
}

.submitConsents {
  margin-top: variables.$base-spacing;
}
