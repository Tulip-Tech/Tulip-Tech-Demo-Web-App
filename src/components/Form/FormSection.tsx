import React, { ReactElement, ReactNode, useCallback, useContext } from 'react';

import { FormContext } from './Form';
import styles from './Form.module.scss';

import Button from '#components/Button/Button';
import LoadingOverlay from '#components/LoadingOverlay/LoadingOverlay';
import useOpaqueId from '#src/hooks/useOpaqueId';
import type { GenericFormValues } from '#types/form';

export interface FormSectionContentArgs<T extends GenericFormValues, TErrors> {
  values: T;
  isEditing: boolean;
  isBusy: boolean;
  onChange: React.ChangeEventHandler<HTMLInputElement | HTMLTextAreaElement>;
  errors?: TErrors | undefined;
}

export interface FormSectionProps<TData extends GenericFormValues, TErrors> {
  className?: string;
  panelHeaderClassName?: string;
  label: string;
  editButton?: string | ReactElement;
  saveButton?: string;
  cancelButton?: string;
  canSave?: (values: TData) => boolean;
  onSubmit?: (values: TData) => Promise<{ errors?: string[] }>;
  content: (args: FormSectionContentArgs<TData, TErrors>) => ReactNode;
  children?: never;
  readOnly?: boolean;
}

export function FormSection<TData extends GenericFormValues>({
  className,
  panelHeaderClassName,
  label,
  editButton,
  saveButton,
  cancelButton,
  canSave,
  onSubmit,
  content,
  readOnly = false,
}: FormSectionProps<TData, string[]>): ReactElement<FormSectionProps<TData, string[]>> | null {
  const sectionId = useOpaqueId(label);
  const {
    formState: { values, activeSectionId, isDirty, errors: formErrors, isBusy },
    setFormState,
    isLoading,
    onCancel,
  } = useContext(FormContext) as FormContext<TData>;

  const isEditing = sectionId === activeSectionId;

  const onChange = useCallback(
    function onChange({ currentTarget }: React.FormEvent<HTMLInputElement | HTMLTextAreaElement>) {
      if (!currentTarget) return;

      const { name, type } = currentTarget;
      const value = type === 'checkbox' ? (currentTarget as HTMLInputElement).checked : currentTarget.value;

      if (!isEditing) {
        onCancel();
      }

      setFormState((oldState) => {
        const newValues = { ...oldState.values };

        // This logic handles nested names like 'consents.terms'
        name.split('.').reduce((parent, field, index, arr) => {
          if (index === arr.length - 1) {
            parent[field] = value;
          } else {
            parent[field] = { ...parent[field] } || {};
          }

          return parent[field];
        }, newValues);

        return {
          ...oldState,
          values: newValues,
          isDirty: true,
          activeSectionId: sectionId,
        };
      });
    },
    [isEditing, onCancel, sectionId, setFormState],
  );

  const handleSubmit = useCallback(
    async function handleSubmit(event?: React.FormEvent) {
      event && event.preventDefault();

      if (onSubmit) {
        let response: { errors?: string[] };

        try {
          setFormState((s) => {
            return { ...s, isBusy: true };
          });
          response = await onSubmit(values);
        } catch (error: unknown) {
          response = { errors: Array.of(error instanceof Error ? error.message : (error as string)) };
        }

        // Don't leave edit mode if there are errors
        if (response?.errors?.length) {
          setFormState((s) => {
            return {
              ...s,
              errors: response?.errors,
              isBusy: false,
            };
          });

          return;
        }
      }

      setFormState((s) => {
        return {
          ...s,
          activeSectionId: undefined,
          isDirty: false,
          isBusy: false,
        };
      });
    },
    [onSubmit, setFormState, values],
  );

  const onEdit = useCallback(
    function onEdit() {
      if (!isEditing) {
        onCancel();

        setFormState((s) => {
          return {
            ...s,
            activeSectionId: sectionId,
          };
        });
      }
    },
    [isEditing, onCancel, sectionId, setFormState],
  );

  return (
    <div className={className}>
      <div className={panelHeaderClassName}>
        <h3>{label}</h3>
      </div>
      {isBusy && isEditing && <LoadingOverlay transparentBackground />}
      {isEditing ? (
        <form className={styles.flexBox} noValidate onSubmit={(event) => event.preventDefault()}>
          {content({ values, isEditing, isBusy, onChange, errors: formErrors })}
        </form>
      ) : (
        <div className={styles.flexBox}>{content({ values, isEditing, isBusy, onChange })}</div>
      )}
      {(saveButton || editButton || cancelButton) && (
        <div className={styles.controls}>
          {isEditing ? (
            <>
              {saveButton && (
                <Button label={saveButton} type="submit" onClick={handleSubmit} disabled={!isDirty || isLoading || (canSave && !canSave(values))} />
              )}
              {cancelButton && <Button label={cancelButton} type="reset" variant="text" onClick={onCancel} />}
            </>
          ) : (
            !readOnly &&
            editButton &&
            (typeof editButton === 'object' ? (
              (editButton as ReactElement)
            ) : (
              <Button label={editButton as string} type="button" onClick={onEdit} disabled={isLoading} />
            ))
          )}
        </div>
      )}
    </div>
  );
}
