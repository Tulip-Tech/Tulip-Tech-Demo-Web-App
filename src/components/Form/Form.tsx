import React, { createContext, SetStateAction, useCallback, useEffect, useState } from 'react';

import type { FormSectionProps } from './FormSection';
import { FormSection } from './FormSection';

import type { GenericFormValues } from '#types/form';

interface Props<TData> {
  initialValues: TData;
  isLoading?: boolean;
  onReset?: () => void;
  children?: FormSectionProps<TData, string[]>[];
}

export interface FormContext<TData> {
  isLoading?: boolean;
  onCancel: () => void;
  formState: FormState<TData>;
  setFormState: React.Dispatch<SetStateAction<FormState<TData>>>;
}

interface FormState<TData> {
  values: TData;
  activeSectionId: string | undefined;
  isDirty: boolean;
  isBusy: boolean;
  errors?: string[];
}

export const FormContext = createContext<FormContext<GenericFormValues> | undefined>(undefined);

function Form<TData extends GenericFormValues>({ isLoading, initialValues, onReset, children }: Props<TData>): JSX.Element {
  const [state, setState] = useState<FormState<TData>>({
    values: initialValues,
    activeSectionId: undefined,
    isDirty: false,
    isBusy: false,
  });

  const onCancel = useCallback(
    function onCancel() {
      setState((s) => {
        return {
          ...s,
          values: initialValues,
          isDirty: false,
          activeSectionId: undefined,
          errors: undefined,
        };
      });
      onReset && onReset();
    },
    [initialValues, onReset],
  );

  useEffect(() => {
    // Don't overwrite values if a section is being edited, since that would wipe out anything the user has entered
    if (!state.activeSectionId) {
      setState((s) => {
        return {
          ...s,
          values: initialValues,
        };
      });
    }
  }, [initialValues, state.activeSectionId]);

  return (
    <FormContext.Provider
      value={{
        isLoading,
        onCancel,
        formState: state,
        setFormState: setState,
      }}
    >
      {children?.map((props, i) => (
        <FormSection key={i} {...props} />
      ))}
    </FormContext.Provider>
  );
}

export default Form;
