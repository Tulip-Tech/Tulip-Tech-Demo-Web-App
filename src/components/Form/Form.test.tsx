import * as React from 'react';
import { render } from '@testing-library/react';

import Form from './Form';

describe('<Form>', () => {
  const initialValues = { test: 'Testing' };

  it('renders Form', () => {
    const { container } = render(
      <Form initialValues={initialValues}>
        {[
          {
            label: 'Edit This',
            editButton: 'Edit Button',
            saveButton: 'Save Button',
            cancelButton: 'Cancel Button',
            content: (section) => (
              <input
                name="test"
                value={section.values.test || ''}
                onChange={() => {
                  /**/
                }}
              />
            ),
          },
        ]}
      </Form>,
    );
    expect(container).toMatchSnapshot();
  });
});
