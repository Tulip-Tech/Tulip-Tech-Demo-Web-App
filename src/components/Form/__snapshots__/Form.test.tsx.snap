// Vitest Snapshot v1

exports[`<Form> > renders Form 1`] = `
<div>
  <div>
    <div>
      <h3>
        Edit This
      </h3>
    </div>
    <div
      class="_flexBox_1c1c63"
    >
      <input
        name="test"
        value="Testing"
      />
    </div>
    <div
      class="_controls_1c1c63"
    >
      <button
        class="_button_f8f296 _default_f8f296 _outlined_f8f296"
        type="button"
      >
        <span>
          Edit Button
        </span>
      </button>
    </div>
  </div>
</div>
`;
